### ISSUE TITLE
Implement .NET Swift interop support targeting OSX and Apple mobile

### ISSUE THREAD

**Issue by kotlarmilos** [dotnet/runtime#95638]
# Overview

The Swift programming language has a [different ABI](https://github.com/apple/swift/blob/main/docs/ABIStabilityManifesto.md), runtime environment, and object model, making it challenging to call into from the .NET runtime. The goal is to establish .NET Interop support in the ecosystem and address Swift ABI differences. Apple's shift towards Swift highlights the importance of Maui being compatible with Swift libraries and frameworks to remain relevant in the Apple ecosystem. Ideally, the interop should operate with a minimal set of additional wrappers and enable direct calls to various Swift functions. 

This project can be divided into at least three separate areas and therefore potentially three separate workstreams. 

**Runtime:** At the lowest level, ensuring that Swift can be called from .NET without an intermediate step through C or assembly is essential. This task must be implemented across all supported .NET runtimes targeting OSX and Apple mobile platforms.

**Code-Generation:** This layer involves the generation of C# bindings for Swift. The projection tooling consists of importing Swift into .NET, exporting .NET to Swift, and tool distribution.

**End-user scenarios:** At the highest level, it is crucial to address end-user scenarios. For instance, a MAUI-iOS developer should be able to create an iOS app that interacts seamlessly with Swift libraries and frameworks. The primary task at this level includes defining and managing the end-to-end process. This includes prioritizing Swift libraries and showcasing the support for .NET Swift interop.

## User stories
 - https://github.com/dotnet/runtime/issues/93631
 - https://github.com/dotnet/runtime/issues/95633
 - https://github.com/dotnet/runtime/issues/95636

**Comment by ShikiSuen** [comment]
Sorry for the interruption, but macOS is not `OSX` anymore.
The major version of macOS changed since 2020.
It is more appropriate to use the term `macOS` instead.

**Comment by kotlarmilos** [comment]
Thanks for your suggestion. The description mentions `macOS aka OSX`. @marek-safar, please inform us if we should make any changes.

**Comment by marek-safar** [comment]
Mixing macOS and Apple seems odd. What about just naming explicitly macOS and iOS ?

**Comment by jkotas** [comment]
> What about just naming explicitly macOS and iOS ?

macCatalyst and tvOS too

We call these "Apple platforms" in many places in the repo, one example from many https://github.com/dotnet/runtime/blob/867250ce3cdd1a93ec08a937c99f61428b8a1082/src/libraries/System.Private.CoreLib/src/System/OperatingSystem.cs#L245 .

**Comment by ShikiSuen** [comment]
@jkotas These platforms have differences in available frameworks.
AppKit is only available on macOS.

If there is a dotNET equivalent of Swift `#if canImport(XXXKit)` then it should be fine.

**Comment by Noemata** [comment]
Is the .Net 9 Swift interop facility usable on Windows?  I would like to use the same code on both Mac and Windows to integrate with Swift libs on both now that the Swift compiler has reached a feasible level of maturity on Windows.  If so, are there any examples out there that demonstrate integration with C# and Swift interop on Windows?  If not, why is Swift compiler integration not supported on Windows?  Lastly, I am aware that Swift on Windows now provides interoperability with Clang/C++, so I'm not looking at that as an alternative given the advantages of the .Net 9 approach for Swift interop.

**Comment by jkotas** [comment]
> Is the .Net 9 Swift interop facility usable on Windows?

Swift interop is Apple-specific for now to reduce scope of the project. Swift interop is a multi-year project even with this scope reduction. Nothing fundamental prevents us from enabling the swift interop on other OSes in future.

**Comment by Noemata** [comment]
@jkotas , "Swift interop is Apple-specific for now to reduce scope of the project." So is that a "no", for Swift interop on Windows with .Net 9?  I don't know enough about how Swift manifests like capabilities on Windows, but given where Swift is today in terms of code "safety", it does appear to be one of the best choices for cross platform low level lib development.  Thus, I would think Swift compiler interop support on Windows should have equal if not higher priority to interop on Apple's eco system, especially given this further elevates C# viability across OS variants.  Please clarify when and if we can expect Swift interop support on Windows.  Please don't cripple us on Windows!

**Comment by jkotas** [comment]
> Swift interop is Apple-specific for now to reduce scope of the project." So is that a "no", for Swift interop on Windows with .Net 9?

Right. Note that Swift interop support in .NET 9 is very limited and most of it is experimental, but it is far from complete and usable for most real-world APIs.

> I would think Swift compiler interop support on Windows should have equal if not higher priority to interop on Apple's eco system,

I disagree. Swift is by several orders of magnitude more popular in Apple eco system than anywhere else. It is why we have prioritized Apple support. We (the core .NET runtime team) do not have infinite resources that forces us to do prioritization decisions like this one.

> Please clarify when and if we can expect Swift interop support on Windows

It is not going to be in .NET 10 and we do not have detailed plans beyond that.

.NET runtime is open-source project that accepts contributions. If you care about Swift interop on Windows a lot, you may consider contributing the implementation (https://github.com/dotnet/runtime/blob/main/CONTRIBUTING.md#dos-and-donts are our contribution guidelines).

**Comment by Noemata** [comment]
@jkotas, agree, Swift is clearly more popular on Apple for obvious reasons.  However, this tendency of Microsoft to quarantine its own platform isn't helping the cause.  It's getting excessively easy to feel like an orphan being a developer on Windows.  WPF / Xamarin / UWP / WinUI / MAUI is not a happy evolution to date.  Enterprise clients in particular are being forced to look beyond Windows as a result.  Thank you for the quick reply.  Please do reconsider your stance.

**Comment by PaperPrototype** [comment]
looking forward to this

**Comment by PaperPrototype** [comment]
Has this issue been resolved? It looks like work has been done on this in https://github.com/dotnet/runtime/issues/95636

**Comment by PaperPrototype** [comment]
This as well https://github.com/dotnet/runtime/issues/108662

**Comment by PaperPrototype** [comment]
Nevermind it looks like there is still work to be done https://github.com/dotnet/runtimelab/issues/3071
